ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ContractsReader.scala

```diff
@@ -6,47 +6,147 @@ package com.daml.platform.store.appendonlydao.events
 import java.time.Instant
 
-import anorm.SqlParser.{binaryStream, int, str}
-import anorm.{Row, RowParser, SimpleSql, SqlParser, SqlStringInterpolation}
+import anorm.SqlParser._
+import anorm.{RowParser, SqlParser, SqlStringInterpolation, _}
 import com.codahale.metrics.Timer
-import com.daml.ledger.participant.state.index.v2.ContractStore
+import com.daml.lf.transaction.GlobalKey
 import com.daml.logging.LoggingContext
 import com.daml.metrics.{Metrics, Timed}
 import com.daml.platform.store.Conversions._
 import com.daml.platform.store.DbType
-import com.daml.platform.store.appendonlydao.DbDispatcher
+import com.daml.platform.store.appendonlydao.LedgerDaoContractsReader._
+import com.daml.platform.store.appendonlydao.events.ContractsReader._
 import com.daml.platform.store.appendonlydao.events.SqlFunctions.{
   H2SqlFunctions,
   PostgresSqlFunctions,
 }
+import com.daml.platform.store.appendonlydao.{DbDispatcher, LedgerDaoContractsReader}
 import com.daml.platform.store.serialization.{Compression, ValueSerializer}
 
 import scala.concurrent.{ExecutionContext, Future}
 
-/** @see [[ContractsTable]]
-  */
 private[appendonlydao] sealed class ContractsReader(
+    val committedContracts: PostCommitValidationData,
     dispatcher: DbDispatcher,
     metrics: Metrics,
-    lfValueTranslationCache: LfValueTranslation.Cache,
     sqlFunctions: SqlFunctions,
 )(implicit ec: ExecutionContext)
-    extends ContractStore {
+    extends LedgerDaoContractsReader {
 
-  val committedContracts: PostCommitValidationData = ContractsTable
-
-  import ContractsReader._
+  override def lookupMaximumLedgerTime(ids: Set[ContractId])(implicit
+      loggingContext: LoggingContext
+  ): Future[Option[Instant]] =
+    Timed.future(
+      metrics.daml.index.db.lookupMaximumLedgerTime,
+      dispatcher
+        .executeSql(metrics.daml.index.db.lookupMaximumLedgerTimeDbMetrics) { implicit connection =>
+          committedContracts.lookupMaximumLedgerTime(ids)
+        }
+        .map(_.get),
+    )
 
-  private val contractRowParser: RowParser[(String, InputStream, Option[Int])] =
-    str("template_id") ~ binaryStream("create_argument") ~ int(
-      "create_argument_compression"
-    ).? map SqlParser.flatten
+  /** Lookup a contract key state at a specific ledger offset.
+    *
+    * @param key the contract key
+    * @param validAt the event_sequential_id of the ledger at which to query for the key state
+    * @return the key state.
+    */
+  override def lookupKeyState(key: GlobalKey, validAt: Long)(implicit
+      loggingContext: LoggingContext
+  ): Future[KeyState] =
+    Timed.future(
+      metrics.daml.index.db.lookupKeyState,
+      dispatcher.executeSql(metrics.daml.index.db.lookupContractByKey) { implicit connection =>
+        SQL"""
+          WITH last_contract_key_create AS (
+                 SELECT contract_id, flat_event_witnesses
+                   FROM participant_events
+                  WHERE event_kind = #$EventKindCreated
```

Comment:
I actually don't mind putting the `event_kind` IDs directly into the query, as it discourages from a "quick refactoring" that changes the numbers.
There could be other alternatives as well. For example: use SQL functions that simply return the respective value, putting it under the governance of migrations.

Above all, we should be consistent in using either hardcoded event kind ids or use interpolation.

Reply:
Done